google-cloud-storage/src/test/java/com/google/cloud/storage/DefaultBufferedWritableByteChannelTest.java

```diff
@@ -0,0 +1,549 @@
+    ByteBuffer data2 = DataGenerator.base64Characters().genByteBuffer(9);
+    ByteBuffer data3 = DataGenerator.base64Characters().genByteBuffer(3);
+    ByteBuffer data4 = DataGenerator.base64Characters().genByteBuffer(10);
+
+    ImmutableList<ByteBuffer> buffers = ImmutableList.of(data1, data2, data3, data4);
+
+    int allDataSize = buffers.stream().mapToInt(ByteBuffer::remaining).sum();
+    byte[] allData =
+        buffers.stream().reduce(ByteBuffer.allocate(allDataSize), ByteBuffer::put).array();
+    buffers.forEach(b -> b.position(0));
+
+    try (ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        CountingWritableByteChannelAdapter adapter =
+            new CountingWritableByteChannelAdapter(Channels.newChannel(baos));
+        BufferedWritableByteChannel c = new DefaultBufferedWritableByteChannel(buffer, adapter)) {
+
+      c.write(data1);
+      adapter.nextWriteMaxConsumptionLimit = 2L;
+      c.write(data2);
+
+      c.write(data3);
+      adapter.nextWriteMaxConsumptionLimit = 7L;
+      c.write(data4);
+
+      c.close();
+      assertThrows(ClosedChannelException.class, () -> c.write(null));
+
+      assertWithMessage("Unexpected total flushed length")
+          .that(adapter.writeEndPoints)
+          .isEqualTo(ImmutableList.of(2L, 12L, 19L, 27L));
+      assertThat(baos.toByteArray()).isEqualTo(allData);
+    }
+  }
+
+  /**
+   * Ensure manually calling flush works.
+   *
+   * <pre>
+   *           0         12
+   * data:    |-----------|
+   *             3  6  9
+   * writes:  |--|--|--|--|
+   *             3
+   * flush 1: |--|
+   *                6
+   * flush 2:    |--|
+   *                   10
+   * flush 3:      |----|
+   *                     12
+   * flush 4:           |-|
+   * </pre>
+   */
+  @Example
+  void manualFlushingIsAccurate() throws IOException {
+    ByteBuffer buffer = ByteBuffer.allocate(5);
+
+    ByteBuffer data1 = DataGenerator.base64Characters().genByteBuffer(3);
+    ByteBuffer data2 = DataGenerator.base64Characters().genByteBuffer(3);
+    ByteBuffer data3 = DataGenerator.base64Characters().genByteBuffer(3);
+    ByteBuffer data4 = DataGenerator.base64Characters().genByteBuffer(3);
+
+    ImmutableList<ByteBuffer> buffers = ImmutableList.of(data1, data2, data3, data4);
+
+    int allDataSize = buffers.stream().mapToInt(ByteBuffer::remaining).sum();
+    byte[] allData =
+        buffers.stream().reduce(ByteBuffer.allocate(allDataSize), ByteBuffer::put).array();
+    buffers.forEach(b -> b.position(0));
+
+    try (ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        CountingWritableByteChannelAdapter adapter =
+            new CountingWritableByteChannelAdapter(Channels.newChannel(baos));
+        BufferedWritableByteChannel c = new DefaultBufferedWritableByteChannel(buffer, adapter)) {
+
+      c.write(data1);
+      c.flush();
+
+      c.write(data2);
+      adapter.nextWriteMaxConsumptionLimit = 2L;
```

Comment:
IIUC this is telling the implementation to flush on the next write because data2 (3) is larger than 2 bytes?

Reply:
Close, this is actually limiting how many bytes will be consumed when the flush happens. The flush threshold itself is still determined based upon the size of the buffer, in this case 5.

I've added several comments around these operations to hopefully help follow the flow.